tool/internal/ast/shared.go

```diff
@@ -8,16 +8,52 @@ import (
 	"go/token"
 
 	"github.com/dave/dst"
+	"github.com/open-telemetry/opentelemetry-go-compile-instrumentation/tool/util"
 )
 
-func ListFuncDecls(file string) ([]*dst.FuncDecl, error) {
-	// Parse the file to get only all the function declarations
-	// So we can use fast variant of AST parsing
-	parser := NewAstParser()
-	root, err := parser.ParseFileFast(file)
+func findFuncDecls(root *dst.File, lambda func(*dst.FuncDecl) bool) ([]*dst.FuncDecl, error) {
+	funcDecls, err := ListFuncDecls(root)
 	if err != nil {
 		return nil, err
 	}
+	// The function with receiver and the function without receiver may have
+	// the same name, so they need to be classified into the same name
+	found := make([]*dst.FuncDecl, 0)
+	for _, funcDecl := range funcDecls {
+		if lambda(funcDecl) {
+			found = append(found, funcDecl)
+		}
+	}
+	return found, nil
+}
+
+func FindFuncDecl(root *dst.File, funcName string) (*dst.FuncDecl, error) {
+	decls, err := findFuncDecls(root, func(funcDecl *dst.FuncDecl) bool {
+		return funcDecl.Name.Name == funcName && !HasReceiver(funcDecl)
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(decls) == 0 {
+		//nolint:nilnil // no function declaration found is not an error
+		return nil, nil
+	}
+	return decls[0], nil
+}
+
+func FindFuncDeclWithRecv(root *dst.File, funcName string) ([]*dst.FuncDecl, error) {
```

Comment:
The current name is misleading IMO: it returns both free functions and methods. Maybe rename it to:
```suggestion
func FindFuncDecl(root *dst.File, funcName string) ([]*dst.FuncDecl, error) {
```